cardtel/utils/playing.py

```diff
@@ -1,25 +1,73 @@
+import deuces
+import gameplay
+from django.db import transaction
+
+def play_cards(player, cards):
+    '''|player| attempts to play |cards| on the table.'''
+    if check_valid_play(player, cards):
+        for card in cards:
+            play_card(player, card)
+    else:
+        raise IllegalPlay
+
+@transaction.atomic
 def play_card(player, card):
-    raise NotImplementedError
+    '''|player| plays |card| on the table.'''
+    player.cards.remove(card)
+    player.game.table.add(card)
 
 def fold(player):
-    raise NotImplementedError
+    '''|player| folds, giving up his ability to play again this point.'''
+    player.has_folded = True
+    player.save()
 
-def check_valid_play(player, card):
-    raise NotImplementedError
+def check_valid_play(player, new_cards):
+    '''Check if |new_cards| makes the table better during |player|'s turn.'''
+    table_deuces_cards = get_deuces_cards(player.game.table.cards.all())
+    new_deuces_cards = get_deuces_cards(new_cards)
+    all_cards = new_deuces_cards + table_deuces_cards
+    current_score = player.game.table.best_hand_score
+    updated_score = evaluate_cards(all_cards)
 
-def update_best_hand(table):
-    # This will be stored on the Table
-    raise NotImplementedError
+    # Check that every card makes the hand better
+    for index in xrange(len(new_deuces_cards)):
+        tmp_cards = all_cards[:index] + all_cards[index+1:]
+        tmp_score = evaluate_cards(tmp_cards)
+        if tmp_score <= updated_score:
+            return False
 
-def finish_game(game):
-    raise NotImplementedError
+    # Check that the score has been improved
+    if updated_score < current_score:
+        table.best_hand_score = updated_score
+        table.save()
```

Comment:
setting up the cache seems like kind of a whole thing though https://docs.djangoproject.com/en/1.11/topics/cache/